ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc

```diff
@@ -0,0 +1,433 @@
+  .statistics = nullptr };
+
+/**
+ * @brief Constructor for snpe_subplugin.
+ */
+snpe_subplugin::snpe_subplugin ()
+    : tensor_filter_subplugin (), configured (false), model_path (nullptr),
+      snpe_h (nullptr), inputMap_h (nullptr), outputMap_h (nullptr), user_buffers ()
+{
+  gst_tensors_info_init (std::addressof (inputInfo));
+  gst_tensors_info_init (std::addressof (outputInfo));
+}
+
+/**
+ * @brief Destructor for snpe subplugin.
+ */
+snpe_subplugin::~snpe_subplugin ()
+{
+  cleanup ();
+}
+
+/**
+ * @brief Method to get empty object.
+ */
+tensor_filter_subplugin &
+snpe_subplugin::getEmptyInstance ()
+{
+  return *(new snpe_subplugin ());
+}
+
+/**
+ * @brief Method to cleanup snpe subplugin.
+ */
+void
+snpe_subplugin::cleanup ()
+{
+  g_free (model_path);
+  model_path = nullptr;
+
+  if (!configured)
+    return;
+
+  if (inputMap_h)
+    Snpe_UserBufferMap_Delete (inputMap_h);
+
+  if (outputMap_h)
+    Snpe_UserBufferMap_Delete (outputMap_h);
+
+  for (auto &ub : user_buffers)
+    if (ub)
+      Snpe_IUserBuffer_Delete (ub);
+
+  user_buffers.clear ();
+
+  if (snpe_h)
+    Snpe_SNPE_Delete (snpe_h);
+
+  snpe_h = nullptr;
+  inputMap_h = nullptr;
+  outputMap_h = nullptr;
+
+  gst_tensors_info_free (std::addressof (inputInfo));
+  gst_tensors_info_free (std::addressof (outputInfo));
+
+  configured = false;
+}
+
+/**
+ * @brief Method to prepare/configure SNPE instance.
+ */
+void
+snpe_subplugin::configure_instance (const GstTensorFilterProperties *prop)
+{
+  Snpe_ErrorCode_t ret = SNPE_SUCCESS;
+
+  if (configured) {
+    /* Already configured */
+    cleanup ();
+  }
+
+  // Log SNPE version
+  Snpe_DlVersion_Handle_t lib_version_h = Snpe_Util_GetLibraryVersion ();
+  nns_logi ("SNPE Version: %s", Snpe_DlVersion_ToString (lib_version_h));
+  Snpe_DlVersion_Delete (lib_version_h);
+
+  // Set Runtime
+  /** @todo support other runtimes */
+  Snpe_Runtime_t runtime = SNPE_RUNTIME_CPU;
+  std::string runtime_str (Snpe_RuntimeList_RuntimeToString (runtime));
+  if (Snpe_Util_IsRuntimeAvailable (runtime) == 0) {
+    cleanup ();
+    throw std::runtime_error ("Runtime " + runtime_str + " is not available");
+  }
+
+  Snpe_RuntimeList_Handle_t runtime_list_h = Snpe_RuntimeList_Create ();
+  if (Snpe_RuntimeList_Add (runtime_list_h, runtime) != SNPE_SUCCESS) {
+    Snpe_RuntimeList_Delete (runtime_list_h);
+    cleanup ();
+    throw std::runtime_error ("Failed to add " + runtime_str);
+  }
+
+  // Load network (dlc file)
+  if (!g_file_test (prop->model_files[0], G_FILE_TEST_IS_REGULAR)) {
+    const std::string err_msg
+        = "Given file " + (std::string) prop->model_files[0] + " is not valid";
+    cleanup ();
+    throw std::invalid_argument (err_msg);
+  }
+
+  model_path = g_strdup (prop->model_files[0]);
+  Snpe_DlContainer_Handle_t container_h = Snpe_DlContainer_Open (model_path);
+  if (!container_h) {
+    cleanup ();
+    throw std::runtime_error ("Failed to open the model file " + std::string (model_path));
```

Comment:
To prevent svace issue later, we need to clear allocated handles before exception.
```
  Snpe_DlContainer_Delete (container_h);
  Snpe_RuntimeList_Delete (runtime_list_h);
  Snpe_SNPEBuilder_Delete (snpebuilder_h);
```

Reply:
I updated the code PTAL